In [153]:
# importing necessary packages
import warnings
warnings.filterwarnings('ignore')

# import pandas and numpy
import pandas as pd
#increase max row and column display value
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 200)
pd.set_option('display.width', 1000)

import numpy as np
# Importing matplotlib and seaborn
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [154]:
# reading data dictionary
data_dictionary = pd.read_csv('data_dictionary.csv')
data_dictionary.head(100)

Acronyms                                        Description
0     CIRCLE_ID  Telecom circle area to which the customer belo...
1           LOC            Local calls  within same telecom circle
2           STD              STD calls  outside the calling circle
3            IC                                     Incoming calls
4            OG                                     Outgoing calls
5           T2T  Operator T to T ie within same operator mobile...
6       T2M                    Operator T to other operator mobile
7       T2O                Operator T to other operator fixed line
8       T2F                         Operator T to fixed lines of T
9       T2C                      Operator T to its own call center
10     ARPU                               Average revenue per user
11      MOU                          Minutes of usage  voice calls
12      AON      Age on network  number of days the customer is...
13     ONNET     All kind of calls within the same operator net...
14   OFFNET       All kind of calls outside the operator T network
15         ROAM  Indicates that customer is in roaming zone dur...
16       SPL                                         Special calls
17      ISD                                              ISD calls
18     RECH                                               Recharge
19      NUM                                                 Number
20      AMT                               Amount in local currency
21      MAX                                                Maximum
22     DATA                                        Mobile internet
23       3G                                              G network
24       AV                                                Average
25      VOL                     Mobile internet usage volume in MB
26       2G                                              G network
27      PCK                  Prepaid service schemes called  PACKS
28    NIGHT         Scheme to use during specific night hours only
29  MONTHLY      Service schemes with validity equivalent to a ...
30    SACHET     Service schemes with validity smaller than a m...
31      *.6                              KPI for the month of June
32      *.7                              KPI for the month of July
33      *.8                            KPI for the month of August
34      FB_USER  Service scheme to avail services of Facebook a...
35      VBC      Volume based cost  when no specific scheme is ...

# Reading and understanding the data

In [155]:
# reading data
telecom_data = pd.read_csv('train.csv')
telecom_data.head()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6   arpu_7   arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  std_og_t2t_mou_7  std_og_t2t_mou_8  std_og_t2m_mou_6  std_og_t2m_mou_7  std_og_t2m_mou_8  std_og_t2f_mou_6  std_og_t2f_mou_7  std_og_t2f_mou_8  std_og_t2c_mou_6  std_og_t2c_mou_7  std_og_t2c_mou_8  std_og_mou_6  std_og_mou_7  std_og_mou_8  isd_og_mou_6  isd_og_mou_7  isd_og_mou_8  spl_og_mou_6  spl_og_mou_7  spl_og_mou_8  og_others_6  og_others_7  og_others_8  total_og_mou_6  total_og_mou_7  \
0   0        109             0.0             0.0             0.0            6/30/2014            7/31/2014            8/31/2014   31.277   87.009    7.527        48.58       124.38         1.29         32.24         96.68          2.33           0.00            0.0            0.0           0.00            0.0           0.00              2.23              0.00              0.28              5.29             16.04              2.33              0.00              0.00              0.00              0.00              0.00              0.00          7.53         16.04          2.61             46.34            124.38              1.01             18.75             80.61               0.0              0.00               0.0               0.0               0.0               0.0               0.0         65.09        204.99          1.01           0.0           0.0           0.0          8.20          0.63          0.00         0.38          0.0          0.0           81.21          221.68   
1   1        109             0.0             0.0             0.0            6/30/2014            7/31/2014            8/31/2014    0.000  122.787   42.953         0.00         0.00         0.00          0.00         25.99         30.89           0.00            0.0            0.0           0.00            0.0           0.00              0.00              0.00              0.00              0.00              0.00              0.00              0.00              0.00              0.00              0.00             22.01             29.79          0.00          0.00          0.00              0.00              0.00              0.00              0.00              0.00               0.0              0.00               0.0               0.0               0.0               0.0               0.0          0.00          0.00          0.00           0.0           0.0           0.0          0.00         30.73         31.66         0.00          0.0          0.0            0.00           30.73   
2   2        109             0.0             0.0             0.0            6/30/2014            7/31/2014            8/31/2014   60.806  103.176    0.000         0.53        15.93         0.00         53.99         82.05          0.00           0.00            0.0            0.0           0.00            0.0           0.00              0.53             12.98              0.00             24.11              0.00              0.00              0.00              0.00              0.00              2.14              0.00              0.00         24.64         12.98          0.00              0.00              2.94              0.00             28.94             82.05               0.0              0.00               0.0               0.0               0.0               0.0               0.0         28.94         84.99          0.00           0.0           0.0           0.0          2.89          1.38          0.00         0.00          0.0          0.0           56.49           99.36   
3  

In [156]:
# check the dimension of dataframe
telecom_data.shape

(69999, 172)

In [157]:
# have a look on length columns
column_list = telecom_data.columns.tolist()
print(len(column_list))

172


In [158]:
# info of dataframe
telecom_data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

- Above indicate some column has object values
- Those columns can be brought to resepective datatype

# Data Cleaning

## Handling null,  missing values, unique values, data type conversions

In [159]:
# Statistical analysis of the numercial features
telecom_data.describe().T

count          mean           std       min         25%        50%         75%        max
id                  69999.0  34999.000000  20207.115084     0.000  17499.5000  34999.000  52498.5000  69998.000
circle_id           69999.0    109.000000      0.000000   109.000    109.0000    109.000    109.0000    109.000
loc_og_t2o_mou      69297.0      0.000000      0.000000     0.000      0.0000      0.000      0.0000      0.000
std_og_t2o_mou      69297.0      0.000000      0.000000     0.000      0.0000      0.000      0.0000      0.000
loc_ic_t2o_mou      69297.0      0.000000      0.000000     0.000      0.0000      0.000      0.0000      0.000
arpu_6              69999.0    283.134365    334.213918 -2258.709     93.5810    197.484    370.7910  27731.088
arpu_7              69999.0    278.185912    344.366927 -1289.715     86.7140    191.588    365.3695  35145.834
arpu_8              69999.0    278.858826    351.924315  -945.808     84.0950    192.234    369.9090  33543.624
onnet_mou_6         67231.0    133.153275    299.963093     0.000      7.4100     34.110    119.3900   7376.710
onnet_mou_7         67312.0    133.894438    311.277193     0.000      6.6750     32.280    115.8375   8157.780
onnet_mou_8         66296.0    132.978257    311.896596     0.000      6.4100     32.100    115.0600  10752.560
offnet_mou_6        67231.0    198.874771    316.818355     0.000     34.8600     96.480    232.9900   8362.360
offnet_mou_7        67312.0    197.153383    322.482226     0.000     32.2400     91.885    227.6300   7043.980
offnet_mou_8        66296.0    196.543577    324.089234     0.000     31.5750     91.800    229.3450  14007.340
roam_ic_mou_6       67231.0      9.765435     57.374429     0.000      0.0000      0.000      0.0000   2850.980
roam_ic_mou_7       67312.0      7.014568     55.960985     0.000      0.0000      0.000      0.0000   4155.830
roam_ic_mou_8       66296.0      7.004892     53.408135     0.000      0.0000      0.000      0.0000   4169.810
roam_og_mou_6       67231.0     14.186457     73.469261     0.000      0.0000      0.000      0.0000   3775.110
roam_og_mou_7       67312.0      9.842191     58.511894     0.000      0.0000      0.000      0.0000   2812.040
roam_og_mou_8       66296.0      9.771783     64.618388     0.000      0.0000      0.000      0.0000   5337.040
loc_og_t2t_mou_6    67231.0     46.904854    150.971758     0.000      1.6600     11.910     40.7400   6431.330
loc_og_t2t_mou_7    67312.0     46.166503    154.739002     0.000      1.6500     11.580     39.7600   7400.660
loc_og_t2t_mou_8    66296.0     45.686109    153.716880     0.000      1.6100     11.740     39.8950  10752.560
loc_og_t2m_mou_6    67231.0     93.238231    162.046699     0.000      9.9200     41.030    110.4300   4696.830
loc_og_t2m_mou_7    67312.0     90.799240    153.852597     0.000     10.0900     40.170    107.5400   4557.140
loc_og_t2m_mou_8    66296.0     91.121447    152.997805     0.000      9.8300     40.350    109.2450   4961.330
loc_og_t2f_mou_6    67231.0      3.743179     13.319542     0.000      0.0000      0.000      2.0600    617.580
loc_og_t2f_mou_7    67312.0      3.777031     13.568110     0.000      0.0000      0.000      2.0800    815.330
loc_og_t2f_mou_8    66296.0      3.661652     13.009193     0.000      0.0000      0.000      2.0300    588.290
loc_og_t2c_mou_6    67231.0      1.126025      5.741811     0.000      0.0000      0.000      0.0000    342.860
loc_og_t2c_mou_7    67312.0      1.361052      7.914113     0.000      0.0000      0.000      0.0000    916.240
loc_og_t2c_mou_8    66296.0      1.420840      6.542202     0.000      0.0000      0.000      0.0000    351.830
loc_og_mou_6        67231.0    143.893585    252.034597     0.000     17.2350     65.190    167.8800  10643.380
loc_og_mou_7        67312.0    140.750120    246.313148     0.000     17.5900     63.430    163.9325   7674.780
loc_og_mou_8        66296.0    140.476486    245.342359     0.000     17.2375     63.520    165.6150  

### Handling unique values

In [160]:
# lets check the columns unique values and drop such columns with its value as 1
unique_1_col=[]
for i in telecom_data.columns:
    if telecom_data[i].nunique() == 1:
        unique_1_col.append(i)
    else:
        pass

telecom_data.drop(unique_1_col, axis=1, inplace = True)
print("\n The following Columns are dropped from the dataset as their unique value is 1. (i.e.)It has no variance in the model\n",
      unique_1_col)



 The following Columns are dropped from the dataset as their unique value is 1. (i.e.)It has no variance in the model
 ['circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou', 'loc_ic_t2o_mou', 'last_date_of_month_6', 'last_date_of_month_7', 'last_date_of_month_8', 'std_og_t2c_mou_6', 'std_og_t2c_mou_7', 'std_og_t2c_mou_8', 'std_ic_t2o_mou_6', 'std_ic_t2o_mou_7', 'std_ic_t2o_mou_8']


In [161]:
# The curent dimensions of the dataset
telecom_data.shape

(69999, 159)

In [162]:
# Checkng the overall missing values in the dataset
((telecom_data.isnull().sum()/telecom_data.shape[0])*100).round(2).sort_values(ascending=False)

date_of_last_rech_data_6    74.90
total_rech_data_6           74.90
arpu_3g_6                   74.90
max_rech_data_6             74.90
count_rech_2g_6             74.90
fb_user_6                   74.90
night_pck_user_6            74.90
arpu_2g_6                   74.90
av_rech_amt_data_6          74.90
count_rech_3g_6             74.90
date_of_last_rech_data_7    74.48
fb_user_7                   74.48
arpu_2g_7                   74.48
arpu_3g_7                   74.48
total_rech_data_7           74.48
count_rech_2g_7             74.48
av_rech_amt_data_7          74.48
count_rech_3g_7             74.48
max_rech_data_7             74.48
night_pck_user_7            74.48
night_pck_user_8            73.69
fb_user_8                   73.69
count_rech_3g_8             73.69
av_rech_amt_data_8          73.69
max_rech_data_8             73.69
count_rech_2g_8             73.69
total_rech_data_8           73.69
arpu_3g_8                   73.69
arpu_2g_8                   73.69
date_of_last_r

- As we can see that the columns with datetime values represented as object, they can be converted into datetime format

In [163]:
# selecting all the columns with datetime format
date_col= telecom_data.select_dtypes(include=['object'])
# Converting the selected columns to datetime format
for i in date_col.columns:
    telecom_data[i] = pd.to_datetime(telecom_data[i])

# Current dimension of the dataset
telecom_data.shape

(69999, 159)

In [164]:
# confirming the conversion of dtype
telecom_data.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 159 columns):
 #    Column                    Dtype         
---   ------                    -----         
 0    id                        int64         
 1    arpu_6                    float64       
 2    arpu_7                    float64       
 3    arpu_8                    float64       
 4    onnet_mou_6               float64       
 5    onnet_mou_7               float64       
 6    onnet_mou_8               float64       
 7    offnet_mou_6              float64       
 8    offnet_mou_7              float64       
 9    offnet_mou_8              float64       
 10   roam_ic_mou_6             float64       
 11   roam_ic_mou_7             float64       
 12   roam_ic_mou_8             float64       
 13   roam_og_mou_6             float64       
 14   roam_og_mou_7             float64       
 15   roam_og_mou_8             float64       
 16   loc_og_t2t_mou_6          float64     

### Handling missing values wrt data recharge

In [165]:
# Handling missing values with respect to `data recharge` attributes
telecom_data[['date_of_last_rech_data_6','total_rech_data_6','max_rech_data_6']].head(10)

date_of_last_rech_data_6  total_rech_data_6  max_rech_data_6
0                      NaT                NaN              NaN
1                      NaT                NaN              NaN
2                      NaT                NaN              NaN
3                      NaT                NaN              NaN
4               2014-06-25                7.0             25.0
5                      NaT                NaN              NaN
6                      NaT                NaN              NaN
7               2014-06-20                1.0            198.0
8                      NaT                NaN              NaN
9                      NaT                NaN              NaN

- Let us conside the column date_of_last_rech_data indicating the date of the last recharge made in any given month for mobile internet. Here it can deduced if the total_rech_data and the max_rech_data also has missing values, the missing values in all the columns mentioned can be considered as meaningful missing.
- Hence imputing 0 as their values.
- Meaningfull missing in this case represents the the customer has not done any recharge for mobile interenet.

### Handling the missing values for the attributes total_rech_data_*, max_rech_data_* and for month 6,7,8

In [166]:
# Code for conditional imputation
for i in range(len(telecom_data)):
  # Handling 'total_rech_data', 'max_rech_data' and for month 6
    if pd.isnull((telecom_data['total_rech_data_6'][i]) and (telecom_data['max_rech_data_6'][i])):
        if pd.isnull(telecom_data['date_of_last_rech_data_6'][i]):
            telecom_data['total_rech_data_6'][i]=0
            telecom_data['max_rech_data_6'][i]=0

  # Handling 'total_rech_data', 'max_rech_data' and for month 7
    if pd.isnull((telecom_data['total_rech_data_7'][i]) and (telecom_data['max_rech_data_7'][i])):
        if pd.isnull(telecom_data['date_of_last_rech_data_7'][i]):
            telecom_data['total_rech_data_7'][i]=0
            telecom_data['max_rech_data_7'][i]=0

  # Handling 'total_rech_data', 'max_rech_data' and for month 8
    if pd.isnull((telecom_data['total_rech_data_8'][i]) and (telecom_data['max_rech_data_8'][i])):
        if pd.isnull(telecom_data['date_of_last_rech_data_8'][i]):
            telecom_data['total_rech_data_8'][i]=0
            telecom_data['max_rech_data_8'][i]=0

### Handling the missing values for the attributes count_rech_2g_*,count_rech_3g_* for month 6,7,8

In [167]:
# Checking the related columns values
telecom_data[['count_rech_2g_6','count_rech_3g_6','total_rech_data_6']].head(10)

count_rech_2g_6  count_rech_3g_6  total_rech_data_6
0              NaN              NaN                0.0
1              NaN              NaN                0.0
2              NaN              NaN                0.0
3              NaN              NaN                0.0
4              7.0              0.0                7.0
5              NaN              NaN                0.0
6              NaN              NaN                0.0
7              1.0              0.0                1.0
8              NaN              NaN                0.0
9              NaN              NaN                0.0

- From the above tablular the column values of total_rech_data for each month from 6 to 8 respectively is the sum of the columns values of count_rech_2g for each month from 6 to 8 respectively and count_rech_3g for each month from 6 to 8 respectively, which derives to a multicollinearity issue. In order to reduce the multicollinearity, we can drop the columns count_rech_2g for each month from 6 to 8 respectively and count_rech_3g for each month from 6 to 8 respectively.

In [168]:
# Dropping the columns 'count_rech_2g_*' & 'count_rech_3g_*' for the months 6,7,8
telecom_data.drop(['count_rech_2g_6','count_rech_3g_6',
                   'count_rech_2g_7','count_rech_3g_7',
                   'count_rech_2g_8','count_rech_3g_8'],axis=1, inplace=True)

In [169]:
# The curent dimensions of the dataset
telecom_data.shape

(69999, 153)

- Handling the missing values for the attributes arpu_3g_*,arpu_2g_* for month 6,7,8

In [170]:
# Checking the related columns values
telecom_data[['arpu_3g_6','arpu_2g_6','av_rech_amt_data_6']].head(10)

arpu_3g_6  arpu_2g_6  av_rech_amt_data_6
0        NaN        NaN                 NaN
1        NaN        NaN                 NaN
2        NaN        NaN                 NaN
3        NaN        NaN                 NaN
4        0.0        0.0               175.0
5        NaN        NaN                 NaN
6        NaN        NaN                 NaN
7        0.0        0.0               198.0
8        NaN        NaN                 NaN
9        NaN        NaN                 NaN

In [171]:
# Checking the correlation between the above mentioned columns in tabular for months 6,7,8 and 9
print("Correlation table for month 6\n\n", telecom_data[['arpu_3g_6','arpu_2g_6','av_rech_amt_data_6']].corr())
print("\nCorrelation table for month 7\n\n", telecom_data[['arpu_3g_7','arpu_2g_7','av_rech_amt_data_7']].corr())
print("\nCorrelation table for month 8\n\n", telecom_data[['arpu_3g_8','arpu_2g_8','av_rech_amt_data_8']].corr())

Correlation table for month 6

                     arpu_3g_6  arpu_2g_6  av_rech_amt_data_6
arpu_3g_6            1.000000   0.930150            0.805494
arpu_2g_6            0.930150   1.000000            0.833773
av_rech_amt_data_6   0.805494   0.833773            1.000000

Correlation table for month 7

                     arpu_3g_7  arpu_2g_7  av_rech_amt_data_7
arpu_3g_7            1.000000    0.93443            0.799627
arpu_2g_7            0.934430    1.00000            0.815300
av_rech_amt_data_7   0.799627    0.81530            1.000000

Correlation table for month 8

                     arpu_3g_8  arpu_2g_8  av_rech_amt_data_8
arpu_3g_8            1.000000   0.923875            0.786412
arpu_2g_8            0.923875   1.000000            0.800717
av_rech_amt_data_8   0.786412   0.800717            1.000000


- From the above correlation table between attributes arpu_2g_* and arpu_3g_* for each month from 6 to 8 respectively is highly correlated to the attribute av_rech_amt_data_* for each month from 6 to 8 respectively.
- Considering the high correlation between them, it is safer to drop the attributes arpu_2g_* and arpu_3g_*.

In [172]:
# Dropping the columns 'arpu_3g_*'&'arpu_2g_*' in month 6,7,8 datafrom the dataset
telecom_data.drop(['arpu_3g_6','arpu_2g_6',
                  'arpu_3g_7','arpu_2g_7',
                  'arpu_3g_8','arpu_2g_8'],axis=1, inplace=True)

In [173]:
# The curent dimensions of the dataset
telecom_data.shape

(69999, 147)

### Handling the other attributes with higher missing value percentage

- The column fb_user_* and night_pck_user_* for each month from 6 to 8 respectively has a missing values above 50% and does not seem to add any information to understand the data. Hence we can drop these columns for further analysis.

In [174]:
telecom_data.drop(['fb_user_6','fb_user_7','fb_user_8',
                  'night_pck_user_6','night_pck_user_7','night_pck_user_8'],
                  axis=1, inplace=True)

In [175]:
# The curent dimensions of the dataset
telecom_data.shape

(69999, 141)

### Handling the missing values for the attributes av_rech_amt_data_* for month 6,7,8

In [176]:
# Checking the related columns values
telecom_data[['av_rech_amt_data_6','max_rech_data_7','total_rech_data_8']].head(10)

av_rech_amt_data_6  max_rech_data_7  total_rech_data_8
0                 NaN              0.0                0.0
1                 NaN            145.0                0.0
2                 NaN              0.0                0.0
3                 NaN              0.0                0.0
4               175.0             41.0                6.0
5                 NaN              0.0                0.0
6                 NaN              0.0                0.0
7               198.0            198.0                1.0
8                 NaN              0.0                0.0
9                 NaN              0.0                0.0

- From the above tabular it is deduced that the missing values for the column av_rech_amt_data_* for each month from 6 to 8 can be replaced as 0 if the total_rech_data_* for each month from 6 to 8 respectively is 0. i.e. if the total recharge done is 0 then the average recharge amount shall also be 0.

In [177]:
# Code for conditional imputation
for i in range(len(telecom_data)):
  # Handling `av_rech_amt_data`  for month 6
    if (pd.isnull(telecom_data['av_rech_amt_data_6'][i]) and (telecom_data['total_rech_data_6'][i]==0)):
        telecom_data['av_rech_amt_data_6'][i] = 0

  # Handling `av_rech_amt_data`  for month 7
    if (pd.isnull(telecom_data['av_rech_amt_data_7'][i]) and (telecom_data['total_rech_data_7'][i]==0)):
        telecom_data['av_rech_amt_data_7'][i] = 0

  # Handling `av_rech_amt_data`  for month 8
    if (pd.isnull(telecom_data['av_rech_amt_data_8'][i]) and (telecom_data['total_rech_data_8'][i]==0)):
        telecom_data['av_rech_amt_data_8'][i] = 0

In [178]:
# Checkng the overall missing values in the dataset
((telecom_data.isnull().sum()/telecom_data.shape[0])*100).round(2).sort_values(ascending=False)

date_of_last_rech_data_6    74.90
date_of_last_rech_data_7    74.48
date_of_last_rech_data_8    73.69
offnet_mou_8                 5.29
roam_ic_mou_8                5.29
roam_og_mou_8                5.29
spl_ic_mou_8                 5.29
loc_og_t2t_mou_8             5.29
ic_others_8                  5.29
loc_og_t2c_mou_8             5.29
loc_og_mou_8                 5.29
loc_og_t2f_mou_8             5.29
loc_og_t2m_mou_8             5.29
isd_ic_mou_8                 5.29
onnet_mou_8                  5.29
loc_ic_t2m_mou_8             5.29
loc_ic_t2t_mou_8             5.29
std_ic_mou_8                 5.29
std_og_mou_8                 5.29
isd_og_mou_8                 5.29
std_ic_t2f_mou_8             5.29
std_og_t2f_mou_8             5.29
std_ic_t2m_mou_8             5.29
std_og_t2m_mou_8             5.29
std_og_t2t_mou_8             5.29
spl_og_mou_8                 5.29
og_others_8                  5.29
std_ic_t2t_mou_8             5.29
loc_ic_t2f_mou_8             5.29
loc_ic_mou_8  

In [179]:
telecom_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 141 entries, id to churn_probability
dtypes: datetime64[ns](6), float64(108), int64(27)
memory usage: 75.3 MB


- From the above results, we can conclude, the date_of_last_rech_data_* corresponding to months 6,7,8 are of no value after the conditional imputation of of columns total_rech_data_*, max_rech_data_*are completes. Also the missing value percentage is high for these columns and can be dropped from the dataset.

In [180]:
# Dropping the columns related to datetime dtype from the dataset
telecom_data.drop(["date_of_last_rech_data_6","date_of_last_rech_data_7",
                   "date_of_last_rech_data_8"], axis=1, inplace=True)

- As we can no more utilise the datetime column, we can drop the date_of_last_rech_data_* column corresponding to months 6,7,8respectively.

In [181]:
# Dropping the columns related to datetime dtype from the dataset
telecom_data.drop(["date_of_last_rech_6","date_of_last_rech_7",
                   "date_of_last_rech_8"], axis=1, inplace=True)

In [182]:
# The curent dimensions of the dataset
telecom_data.shape

(69999, 135)

## Filtering the High Value Customer from Good Phase

In [183]:
# Calculating the total recharge amount done for data alone in months 6, since month 6 is good phase customer
telecom_data['total_rech_amt_data_6']=telecom_data['av_rech_amt_data_6'] * telecom_data['total_rech_data_6']
#telecom_data['total_rech_amt_data_7']=telecom_data['av_rech_amt_data_7'] * telecom_data['total_rech_data_7']

# Calculating the overall recharge amount for the months 6, since month 6  is good phase customer 
telecom_data['overall_rech_amt_6'] = telecom_data['total_rech_amt_data_6'] + telecom_data['total_rech_amt_6']
#telecom_data['overall_rech_amt_7'] = telecom_data['total_rech_amt_data_7'] + telecom_data['total_rech_amt_7']

# Calculating the average recharge done by customer in months June and July(i.e. 6th and 7th month)
#telecom_data['avg_rech_amt_6_7'] = (telecom_data['overall_rech_amt_6'] + telecom_data['overall_rech_amt_7'])/2

# Finding the value of 70th percentage in the overall revenues defining the high value customer creteria for the company
cut_off = telecom_data['overall_rech_amt_6'].quantile(0.70)
print("\nThe 70th quantile value to determine the High Value Customer is: ",cut_off,"\n")

# Filtering the data to the top 30% considered as High Value Customer
telecom_data = telecom_data[telecom_data['overall_rech_amt_6'] >= cut_off]



The 70th quantile value to determine the High Value Customer is:  476.0 



In [184]:
# The curent dimension of the dataset
telecom_data.shape

(21157, 137)

- The total number of customers is now limited to ~20k who lies under the High Value customer criteria basen upon which the model is built.

In [185]:
# Checkng the overall missing values in the dataset
((telecom_data.isnull().sum()/telecom_data.shape[0])*100).round(2).sort_values(ascending=False)

loc_ic_mou_8             6.04
std_ic_t2t_mou_8         6.04
std_ic_t2m_mou_8         6.04
roam_og_mou_8            6.04
loc_og_t2t_mou_8         6.04
loc_ic_t2m_mou_8         6.04
loc_ic_t2f_mou_8         6.04
spl_ic_mou_8             6.04
isd_ic_mou_8             6.04
loc_ic_t2t_mou_8         6.04
loc_og_t2m_mou_8         6.04
loc_og_t2f_mou_8         6.04
loc_og_t2c_mou_8         6.04
isd_og_mou_8             6.04
spl_og_mou_8             6.04
og_others_8              6.04
std_og_t2f_mou_8         6.04
std_og_mou_8             6.04
loc_og_mou_8             6.04
std_og_t2t_mou_8         6.04
std_og_t2m_mou_8         6.04
std_ic_t2f_mou_8         6.04
std_ic_mou_8             6.04
onnet_mou_8              6.04
ic_others_8              6.04
offnet_mou_8             6.04
roam_ic_mou_8            6.04
ic_others_7              3.13
offnet_mou_7             3.13
std_ic_t2f_mou_7         3.13
roam_ic_mou_7            3.13
onnet_mou_7              3.13
std_ic_mou_7             3.13
std_og_mou

- The remaining attributes with missing value can be imputed using the advanced imputation technique like KNNImputer

In [186]:
# Numerical columns available
num_col = telecom_data.select_dtypes(include = ['int64','float64']).columns.tolist()

In [ ]:
# Importing the libraries for Scaling and Imputation
from sklearn.impute import KNNImputer
from sklearn.preprocessing import MinMaxScaler

# Calling the Scaling function
scalar = MinMaxScaler()

# Scaling and transforming the data for the columns that are numerical
telecom_data[num_col]=scalar.fit_transform(telecom_data[num_col])
knn=KNNImputer(n_neighbors=3)

telecom_data_knn = pd.DataFrame(knn.fit_transform(telecom_data[num_col]))
telecom_data_knn.columns=telecom_data[num_col].columns

In [ ]:
# check for any null values after imputation for numerical columns
telecom_data_knn.isnull().sum().sum()

- The KNN Imputer has replaced all the null values in the numerical column usingK-means algorithm sucessfully

In [ ]:
# Since we sclaed the numerical columns for the purpose of handling the null values, 
# we can restore the sclaed alues to its original form.

# Converting the scaled data back to the original data
telecom_data[num_col]=scalar.inverse_transform(telecom_data_knn)

# Checking the top 10 data
telecom_data.head(10)

In [ ]:
# Checking the overall missing values in the dataset
((telecom_data.isnull().sum()/telecom_data.shape[0])*100).round(2).sort_values(ascending=False)


In [ ]:
# lets find out churn/non churn percentage
print((telecom_data['churn_probability'].value_counts()/len(telecom_data))*100)
((telecom_data['churn_probability'].value_counts()/len(telecom_data))*100).plot(kind="pie")
plt.show()

### As we can see that 90% of the customers do not churn, there is a possibility of class imbalance

In [ ]:
# Selecting all the churn phase columns in order to drop then
churn_phase_cols = [col for col in telecom_data.columns if '_8' in col]
print("The columns from churn phase are:\n",churn_phase_cols)

In [ ]:
# Dropping the selected churn phase columns
telecom_data.drop(churn_phase_cols, axis=1, inplace=True)

# The curent dimension of the dataset after dropping the churn related columns
telecom_data.shape

- We can still clean the data by few possible columns relating to the good phase.
- As we derived few columns in the good phase earlier, we can drop those related

In [ ]:
telecom_data.drop(['total_rech_amt_data_6','av_rech_amt_data_6',
                   'total_rech_data_6','total_rech_amt_6'], axis=1, inplace=True)

- We can also create new columns for the defining the good phase variables and drop the seperate 6th and 7 month variables.
- Before proceding to check the remaining missing value handling, let us check the collineartity of the indepedent variables and try to understand their dependencies.

In [ ]:
# creating a list of column names for each month
mon_6_cols = [col for col in telecom_data.columns if '_6' in col]
mon_7_cols = [col for col in telecom_data.columns if '_7' in col]

In [ ]:
# lets check the correlation amongst the independent variables, drop the highly correlated ones
telecom_data_corr = telecom_data.corr()
telecom_data_corr.loc[:,:] = np.tril(telecom_data_corr, k=-1)
telecom_data_corr = telecom_data_corr.stack()
telecom_data_corr
telecom_data_corr[(telecom_data_corr > 0.80) | (telecom_data_corr < -0.80)].sort_values(ascending=False)

In [ ]:
col_to_drop=['total_rech_amt_7', 'isd_og_mou_7', 'sachet_2g_7', 'total_ic_mou_6' , 'total_ic_mou_7' , 
             'std_og_t2t_mou_6']

# These columns can be dropped as they are highly collinered with other predictor variables.
# criteria set is for collinearity of 85%

#  dropping these column
telecom_data.drop(col_to_drop, axis=1, inplace=True)

In [ ]:
telecom_data.shape

## Deriving new variables to understand the data

In [ ]:
# We have a column called 'aon'

# we can derive new variables from this to explain the data w.r.t churn.

# creating a new variable 'tenure'
telecom_data['tenure'] = (telecom_data['aon']/30).round(0)

# Since we derived a new column from 'aon', we can drop it
telecom_data.drop('aon',axis=1, inplace=True)

In [ ]:
# Checking the distribution of he tenure variable

sns.distplot(telecom_data['tenure'],bins=30)
plt.show()

In [ ]:
tn_range = [0, 6, 12, 24, 60, 61]
tn_label = [ '0-6 Months', '6-12 Months', '1-2 Yrs', '2-5 Yrs', '5 Yrs and above']
telecom_data['tenure_range'] = pd.cut(telecom_data['tenure'], tn_range, labels=tn_label)
telecom_data['tenure_range'].head()

In [ ]:
# Plotting a bar plot for tenure range
plt.figure(figsize=[12,7])
sns.barplot(x='tenure_range',y='churn_probability', data=telecom_data)
plt.show()

- It can be seen that the maximum churn rate happens within 0-6 month, but it gradually decreases as the customer retains in the network.
- The average revenue per user is good phase of customer is given by arpu_6.

In [ ]:
# Visualizing the column created
sns.distplot(telecom_data['arpu_6'])
plt.show()

In [ ]:
plt.figure(figsize=(10,50))
telecom_data_corr = telecom_data.loc[:, telecom_data.columns != 'tenure_range'].corr()
heatmap_churn = sns.heatmap(telecom_data_corr.corr()[['churn_probability']].sort_values(ascending=False, by='churn_probability'),annot=True, 
                                cmap='summer')
heatmap_churn.set_title("Features Correlating with Churn variable", fontsize=15)
plt.show()

- creating a scatter plot with total_rech_num_7 and  arpu_7

In [ ]:
# lets now draw a scatter plot between total recharge and avg revenue for the 8th month
telecom_data[['total_rech_num_7', 'arpu_7']].plot.scatter(x = 'total_rech_num_7',
                                                              y='arpu_7')
plt.show()

- creating a box plot with churn_probability and tenure

In [ ]:
sns.boxplot(x = telecom_data.churn_probability, y = telecom_data.tenure)
plt.show()

- From the above plot , its clear tenured customers do no churn and they keep availing telecom servicesm

In [ ]:
# Plot between churn vs max rechare amount
ax = sns.kdeplot(telecom_data.max_rech_amt_7[(telecom_data["churn_probability"] == 0)],
                color="Red", shade = True)
ax = sns.kdeplot(telecom_data.max_rech_amt_7[(telecom_data["churn_probability"] == 1)],
                ax =ax, color="Blue", shade= True)
ax.legend(["No-Churn","Churn"],loc='upper right')
ax.set_ylabel('Density')
ax.set_xlabel('Volume based cost')
ax.set_title('Distribution of Max Recharge Amount by churn')
plt.show()

In [ ]:
# churn vs avg rechare amount
ax = sns.kdeplot(telecom_data.av_rech_amt_data_7[(telecom_data["churn_probability"] == 0)],
                color="Red", shade = True)
ax = sns.kdeplot(telecom_data.av_rech_amt_data_7[(telecom_data["churn_probability"] == 1)],
                ax =ax, color="Blue", shade= True)
ax.legend(["No-Churn","Churn"],loc='upper right')
ax.set_ylabel('Density')
ax.set_xlabel('Volume based cost')
ax.set_title('Distribution of Average Recharge Amount for Data by churn')
plt.show()

- Creating categories for month 7 month column totalrecharge and their count

In [ ]:
# Plotting the results
telecom_data['total_rech_data_group_7'] = pd.cut(
    telecom_data['total_rech_data_7'],
    bins=[-1, 0, 10, 25, 100],
    labels=["No_Recharge", "<=10_Recharges", "10-25_Recharges", ">25_Recharges"]
)

# Create the count plot
plt.figure(figsize=(10, 6))
sns.countplot(x='total_rech_data_group_7', data=telecom_data)
plt.show() 

In [ ]:
 telecom_data['total_rech_num_group_7'] = pd.cut(
    telecom_data['total_rech_num_7'],
    bins=[-1,0,10,25,1000],
    labels=["No_Recharge", "<=10_Recharges", "10-25_Recharges", ">25_Recharges"]
)

# Create the count plot
plt.figure(figsize=(10, 6))
sns.countplot(x='total_rech_num_group_7', data=telecom_data)
plt.show() 


In [ ]:
# Creating a dummy variable for some of the categorical variables and dropping the first one.
dummy = pd.get_dummies(telecom_data[['total_rech_data_group_7','total_rech_num_group_7','tenure_range']], drop_first=True, dtype=int)
dummy.head()

In [ ]:
# Adding the results to the master dataframe
telecom_data = pd.concat([telecom_data, dummy], axis=1)
telecom_data.head()

In [ ]:
# Creating a copy of the filtered dataframe
df = telecom_data[:].copy()
# Dropping unwanted columns
df.drop(['tenure_range','total_rech_data_group_7','total_rech_num_group_7','tenure'], axis=1, inplace=True)

In [ ]:
# Cheking the dataset
df.head()

In [ ]:
# lets create X dataset for model building.
X = df.drop(['churn_probability'],axis=1)

In [ ]:
X.head()

In [ ]:
# lets create y dataset for model building.
y=df['churn_probability']
y.head()

In [ ]:
# split the dateset into train and test datasets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, train_size=0.7, random_state=1)
X_train.shape, X_test.shape

In [ ]:
X_train.info(verbose=True)


In [ ]:
num_col = X_train.select_dtypes(include = ['int64','float64']).columns.tolist()

In [ ]:
# apply scaling on the dataset
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
X_train[num_col] = scaler.fit_transform(X_train[num_col])

In [ ]:
X_train.head()


## Data Imbalance Handling

In [ ]:
from imblearn.over_sampling import SMOTE
sm = SMOTE(random_state=42)
X_train_sm,y_train_sm = sm.fit_resample(X_train,y_train)

In [ ]:
X_train_sm.shape, y_train_sm.shape

## Logistic Regression

In [ ]:
import statsmodels.api as sm
# Logistic regression model
logm1 = sm.GLM(y_train_sm,(sm.add_constant(X_train_sm)), family = sm.families.Binomial())
logm1.fit().summary()

## Logistic Regression using Feature Selection (RFE method)

In [ ]:
from sklearn.linear_model import LogisticRegression
logreg = LogisticRegression()

from sklearn.feature_selection import RFE

# running RFE with 20 variables as output
rfe = RFE(logreg,  n_features_to_select= 20)             
rfe = rfe.fit(X_train, y_train)

In [ ]:
rfe.support_

In [ ]:
rfe_columns=X_train_sm.columns[rfe.support_]
rfe_columns

In [ ]:
list(zip(X_train.columns, rfe.support_, rfe.ranking_))

##### Assessing the model with StatsModels

In [ ]:
X_train_SM = sm.add_constant(X_train_sm[rfe_columns])
logm2 = sm.GLM(y_train_sm,X_train_SM, family = sm.families.Binomial())
res = logm2.fit()
res.summary()

In [ ]:
# Getting the predicted values on the train set
y_train_sm_pred = res.predict(X_train_SM)
y_train_sm_pred = y_train_sm_pred.values.reshape(-1)
y_train_sm_pred[:10]

##### Creating a dataframe with the actual churn flag and the predicted probabilities

In [ ]:
y_train_sm_pred_final = pd.DataFrame({'Converted':y_train_sm.values, 'Converted_prob':y_train_sm_pred})
y_train_sm_pred_final.head()

##### Creating new column 'predicted' with 1 if Churn_Prob > 0.5 else 0

In [ ]:
y_train_sm_pred_final['churn_pred'] = y_train_sm_pred_final.Converted_prob.map(lambda x: 1 if x > 0.5 else 0)

# Viewing the prediction results
y_train_sm_pred_final.head()

In [ ]:
from sklearn import metrics
# Confusion matrix 
confusion = metrics.confusion_matrix(y_train_sm_pred_final.Converted, y_train_sm_pred_final.churn_pred )
print(confusion)

| Attempt/Predicted | not_churn | churn |
| :---: | :---: | :---: 
| not_churn | 8906 | 4346 |
| churn | 2730 | 10522 |

In [ ]:
# Checking the overall accuracy.
print("The overall accuracy of the model is:",metrics.accuracy_score(y_train_sm_pred_final.Converted, y_train_sm_pred_final.churn_pred))

#### Checking VIFs

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Create a dataframe that will contain the names of all the feature variables and their respective VIFs
vif = pd.DataFrame()
vif['Features'] = X_train_sm[rfe_columns].columns
vif['VIF'] = [variance_inflation_factor(X_train_sm[rfe_columns].values, i) for i in range(X_train_sm[rfe_columns].shape[1])]
vif['VIF'] = round(vif['VIF'], 2)
vif = vif.sort_values(by = "VIF", ascending = False)
vif

### Metrics beyond simply accuracy

In [ ]:
TP = confusion[1,1] # true positive 
TN = confusion[0,0] # true negatives
FP = confusion[0,1] # false positives
FN = confusion[1,0] # false negatives

In [ ]:
# Let's see the sensitivity of our logistic regression model
print("Sensitivity = ",TP / float(TP+FN))

# Let us calculate specificity
print("Specificity = ",TN / float(TN+FP))

# Calculate false postive rate - predicting churn when customer does not have churned
print("False Positive Rate = ",FP/ float(TN+FP))

# positive predictive value 
print ("Precision = ",TP / float(TP+FP))

# Negative predictive value
print ("True Negative Prediction Rate = ",TN / float(TN+ FN))

### Plotting the ROC Curve

In [ ]:
# Defining a function to plot the roc curve
def draw_roc( actual, probs ):
    fpr, tpr, thresholds = metrics.roc_curve( actual, probs,
                                              drop_intermediate = False )
    auc_score = metrics.roc_auc_score( actual, probs )
    plt.figure(figsize=(5, 5))
    plt.plot( fpr, tpr, label='ROC curve (area = %0.2f)' % auc_score )
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate or [1 - True Negative Prediction Rate]')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

    return None

In [ ]:
# Defining the variables to plot the curve
fpr, tpr, thresholds = metrics.roc_curve( y_train_sm_pred_final.Converted, y_train_sm_pred_final.Converted_prob, drop_intermediate = False )

In [ ]:
# Plotting the curve for the obtained metrics
draw_roc(y_train_sm_pred_final.Converted, y_train_sm_pred_final.Converted_prob)

### Finding Optimal Cutoff Point

In [ ]:
# Let's create columns with different probability cutoffs 
numbers = [float(x)/10 for x in range(10)]
for i in numbers:
    y_train_sm_pred_final[i]= y_train_sm_pred_final.Converted_prob.map(lambda x: 1 if x > i else 0)
y_train_sm_pred_final.head()

In [ ]:
# Now let's calculate accuracy sensitivity and specificity for various probability cutoffs.
cutoff_df = pd.DataFrame( columns = ['probability','accuracy','sensitivity','specificity'])
from sklearn.metrics import confusion_matrix

# TP = confusion[1,1] # true positive 
# TN = confusion[0,0] # true negatives
# FP = confusion[0,1] # false positives
# FN = confusion[1,0] # false negatives

num = [0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9]
for i in num:
    cm1 = metrics.confusion_matrix(y_train_sm_pred_final.Converted, y_train_sm_pred_final[i] )
    total1=sum(sum(cm1))
    accuracy = (cm1[0,0]+cm1[1,1])/total1
    
    specificity = cm1[0,0]/(cm1[0,0]+cm1[0,1])
    sensitivity = cm1[1,1]/(cm1[1,0]+cm1[1,1])
    cutoff_df.loc[i] =[ i ,accuracy,sensitivity,specificity]
print(cutoff_df)


In [ ]:
# plotting accuracy sensitivity and specificity for various probabilities calculated above.
cutoff_df.plot.line(x='probability', y=['accuracy','sensitivity','specificity'])
plt.show()

- From the above graph, we can see the optimum cutoff is slightly higher than 0.5 but lies lower than 0.6. So lets tweek a little more within this range

In [ ]:
# Let's create columns with refined probability cutoffs 
numbers = [0.50,0.51,0.52,0.53,0.54,0.55,0.56,0.57,0.58,0.59]
for i in numbers:
    y_train_sm_pred_final[i]= y_train_sm_pred_final.Converted_prob.map(lambda x: 1 if x > i else 0)
y_train_sm_pred_final.head()

In [ ]:
# Now let's calculate accuracy sensitivity and specificity for various probability cutoffs.
cutoff_df = pd.DataFrame( columns = ['probability','accuracy','sensitivity','specificity'])
from sklearn.metrics import confusion_matrix

# TP = confusion[1,1] # true positive 
# TN = confusion[0,0] # true negatives
# FP = confusion[0,1] # false positives
# FN = confusion[1,0] # false negatives

num = [0.50,0.51,0.52,0.53,0.54,0.55,0.56,0.57,0.58,0.59]
for i in num:
    cm1 = metrics.confusion_matrix(y_train_sm_pred_final.Converted, y_train_sm_pred_final[i] )
    total1=sum(sum(cm1))
    accuracy = (cm1[0,0]+cm1[1,1])/total1
    
    specificity = cm1[0,0]/(cm1[0,0]+cm1[0,1])
    sensitivity = cm1[1,1]/(cm1[1,0]+cm1[1,1])
    cutoff_df.loc[i] =[ i ,accuracy,sensitivity,specificity]
print(cutoff_df)

In [ ]:
# plotting accuracy sensitivity and specificity for various probabilities calculated above.
cutoff_df.plot.line(x='probability', y=['accuracy','sensitivity','specificity'])
plt.show()

- From the above graph we can conclude, the optimal cutoff point in the probability to define the predicted churn variabe converges at 0.54

In [ ]:
y_train_sm_pred_final['final_churn_pred'] = y_train_sm_pred_final.Converted_prob.map( lambda x: 1 if x > 0.54 else 0)

y_train_sm_pred_final.head()

In [ ]:
# Calculating the ovearall accuracy again
print("The overall accuracy of the model now is:",metrics.accuracy_score(y_train_sm_pred_final.Converted, y_train_sm_pred_final.final_churn_pred))

In [ ]:
confusion2 = metrics.confusion_matrix(y_train_sm_pred_final.Converted, y_train_sm_pred_final.final_churn_pred )
print(confusion2)

In [ ]:
TP2 = confusion2[1,1] # true positive 
TN2 = confusion2[0,0] # true negatives
FP2 = confusion2[0,1] # false positives
FN2 = confusion2[1,0] # false negatives

# Let's see the sensitivity of our logistic regression model
print("Sensitivity = ",TP2 / float(TP2+FN2))

# Let us calculate specificity
print("Specificity = ",TN2 / float(TN2+FP2))

# Calculate false postive rate - predicting churn when customer does not have churned
print("False Positive Rate = ",FP2/ float(TN2+FP2))

# positive predictive value 
print ("Precision = ",TP2 / float(TP2+FP2))

# Negative predictive value
print ("True Negative Prediction Rate = ",TN2 / float(TN2 + FN2))

## Precision and recall tradeoff

In [ ]:
from sklearn.metrics import precision_recall_curve
p, r, thresholds = precision_recall_curve(y_train_sm_pred_final.Converted, y_train_sm_pred_final.Converted_prob)

# Plotting the curve
plt.plot(thresholds, p[:-1], "g-")
plt.plot(thresholds, r[:-1], "r-")
plt.show()

### Making predictions on the test set

In [ ]:
# Scaling the test data
X_test[num_col] = scaler.transform(X_test[num_col])
X_test.head()

In [ ]:
# Feature selection
X_test=X_test[rfe_columns]
X_test.head()

In [ ]:
# Adding constant to the test model.
X_test_SM = sm.add_constant(X_test)

In [ ]:
y_test_pred = res.predict(X_test_SM)
print("\n The first ten probability value of the prediction are:\n",y_test_pred[:10])

In [ ]:
y_pred = pd.DataFrame(y_test_pred)
y_pred.head()

In [ ]:
y_pred=y_pred.rename(columns = {0:"Conv_prob"})

In [ ]:
y_test_df = pd.DataFrame(y_test)
y_test_df.head()

In [ ]:
y_pred_final = pd.concat([y_test_df,y_pred],axis=1)
y_pred_final.head()

In [ ]:
y_pred_final['test_churn_pred'] = y_pred_final.Conv_prob.map(lambda x: 1 if x>0.54 else 0)
y_pred_final.head()

In [ ]:
# Checking the overall accuracy of the predicted set.
metrics.accuracy_score(y_pred_final.churn_probability, y_pred_final.test_churn_pred)

### Metrics Evaluation

In [ ]:
# Confusion Matrix
confusion2_test = metrics.confusion_matrix(y_pred_final.churn_probability, y_pred_final.test_churn_pred)
print("Confusion Matrix\n",confusion2_test)

In [ ]:
# Calculating model validation parameters
TP3 = confusion2_test[1,1] # true positive 
TN3 = confusion2_test[0,0] # true negatives
FP3 = confusion2_test[0,1] # false positives
FN3 = confusion2_test[1,0] # false negatives

In [ ]:
# Let's see the sensitivity of our logistic regression model
print("Sensitivity = ",TP3 / float(TP3+FN3))

# Let us calculate specificity
print("Specificity = ",TN3 / float(TN3+FP3))

# Calculate false postive rate - predicting churn when customer does not have churned
print("False Positive Rate = ",FP3/ float(TN3+FP3))

# positive predictive value 
print ("Precision = ",TP3 / float(TP3+FP3))

# Negative predictive value
print ("True Negative Prediction Rate = ",TN3 / float(TN3+FN3))

### Explaining the results

In [ ]:
print("The accuracy of the predicted model is: ",round(metrics.accuracy_score(y_pred_final.churn_probability, y_pred_final.test_churn_pred),2)*100,"%")
print("The sensitivity of the predicted model is: ",round(TP3 / float(TP3+FN3),2)*100,"%")

print("\nAs the model created is based on a sentivity model, i.e. the True positive rate is given more importance as the actual and prediction of churn by a customer\n") 

In [ ]:
# ROC curve for the test dataset

# Defining the variables to plot the curve
fpr, tpr, thresholds = metrics.roc_curve(y_pred_final.churn_probability,y_pred_final.Conv_prob, drop_intermediate = False )
# Plotting the curve for the obtained metrics
draw_roc(y_pred_final.churn_probability,y_pred_final.Conv_prob)

- The AUC score of 0.81 for the training dataset and 0.78 for the testing dataset indicates that this model can be deemed effective.

# Logistic Regression using PCA

In [ ]:
# split the dateset into train and test datasets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, train_size=0.7, random_state=1)
print("Dimension of X_train:", X_train.shape)
print("Dimension of X_test:", X_test.shape)

# apply scaling on the dataset

scaler = MinMaxScaler()
X_train[num_col] = scaler.fit_transform(X_train[num_col])
X_test[num_col] = scaler.transform(X_test[num_col])

# Applying SMOTE technique for data imbalance correction

sm = SMOTE(random_state=42)
X_train_sm,y_train_sm = sm.fit_resample(X_train,y_train)
print("Dimension of X_train_sm Shape:", X_train_sm.shape)
print("Dimension of y_train_sm Shape:", y_train_sm.shape)

X_train_sm.head()

In [ ]:
# importing PCA
from sklearn.decomposition import PCA
pca = PCA(random_state=42)

# applying PCA on train data
pca.fit(X_train_sm)

In [ ]:
X_train_sm_pca=pca.fit_transform(X_train_sm)
print("Dimension of X_train_sm_pca: ",X_train_sm_pca.shape)

X_test_pca=pca.transform(X_test)
print("Dimension of X_test_pca: ",X_test_pca.shape)

In [ ]:
#Viewing the PCA components
pca.components_

### Performing Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
logreg_pca = LogisticRegression()
logreg_pca.fit(X_train_sm_pca, y_train_sm)

# making the predictions
y_pred = logreg_pca.predict(X_test_pca)

# converting the prediction into a dataframe
y_pred_df = pd.DataFrame(y_pred)
print("Dimension of y_pred_df:", y_pred_df.shape)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score

# Checking the Confusion matrix
print("Confusion Matirx for y_test & y_pred\n",confusion_matrix(y_test,y_pred),"\n")

# Checking the Accuracy of the Predicted model.
print("Accuracy of the logistic regression model with PCA: ",accuracy_score(y_test,y_pred))

In [ ]:
plt.bar(range(1,len(pca.explained_variance_ratio_)+1),pca.explained_variance_ratio_)
plt.show()

In [ ]:
var_cumu = np.cumsum(pca.explained_variance_ratio_)
fig = plt.figure(figsize=[12,8])
plt.vlines(x=15, ymax=1, ymin=0, colors="r", linestyles="--")
plt.hlines(y=0.95, xmax=30, xmin=0, colors="g", linestyles="--")
plt.plot(var_cumu)
plt.ylabel("Cumulative variance explained")
plt.show()

In [ ]:
np.cumsum(np.round(pca.explained_variance_ratio_, decimals=3)*100)

- More than 95% of the data can be explained with 18 PCA components
- Fitting the dataset with the 20 explainable components

In [ ]:
pca_20 = PCA(n_components=18)

train_pca_20 = pca_20.fit_transform(X_train_sm)
print("Dimension for Train dataset using PCA: ", train_pca_20.shape)

test_pca_20 = pca_20.transform(X_test)
print("Dimension for Test dataset using PCA: ", test_pca_20.shape)

In [ ]:
logreg_pca_20 = LogisticRegression()
logreg_pca_20.fit(train_pca_20, y_train_sm)

# making the predictions
y_pred_20 = logreg_pca_20.predict(test_pca_20)

# converting the prediction into a dataframe
y_pred_df_20 = pd.DataFrame(y_pred_20)
print("Dimension of y_pred_df_20: ", y_pred_df_20.shape)

In [ ]:
# Checking the Confusion matrix
print("Confusion Matirx for y_test & y_pred\n",confusion_matrix(y_test,y_pred_20),"\n")

# Checking the Accuracy of the Predicted model.
print("Accuracy of the logistic regression model with PCA: ",accuracy_score(y_test,y_pred_20))

# Decision Tree with PCA

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
from IPython.display import Image  
from six import StringIO  
from sklearn.tree import export_graphviz
import pydotplus
import graphviz

In [ ]:
dt = DecisionTreeClassifier(random_state=42)

In [ ]:
params = {
    'max_depth': [2, 3, 5, 10, 20],
    'min_samples_leaf': [5, 10, 20, 50, 100],
    'min_samples_split': [50, 150, 50]
}

In [ ]:
# Instantiate the grid search model
grid_search = GridSearchCV(estimator=dt, 
                           param_grid=params, 
                           cv=4, n_jobs=-1, verbose=1, scoring = "accuracy")

In [ ]:
grid_search.fit(X_train_sm_pca, y_train_sm)

In [ ]:
score_df = pd.DataFrame(grid_search.cv_results_)
score_df.head()

In [ ]:
score_df.nlargest(5,"mean_test_score")

In [ ]:
grid_search.best_estimator_

In [ ]:
dt_best = DecisionTreeClassifier( random_state = 42,
                                  max_depth=20, 
                                  min_samples_leaf=5,
                                  min_samples_split=50)

In [ ]:
dt_best.fit(X_train_sm_pca, y_train_sm)

In [ ]:
def evaluate_model(dt_classifier):
    print("Train Accuracy :", accuracy_score(y_train_sm, dt_classifier.predict(X_train_sm_pca)))
    print("Train Confusion Matrix:")
    print(confusion_matrix(y_train_sm, dt_classifier.predict(X_train_sm_pca)))
    print("-"*50)
    print("Test Accuracy :", accuracy_score(y_test, dt_classifier.predict(X_test_pca)))
    print("Test Confusion Matrix:")
    print(confusion_matrix(y_test, dt_classifier.predict(X_test_pca)))

In [ ]:
evaluate_model(dt_best)

## Random Forest with PCA

In [ ]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=10, max_depth=4, max_features=5, random_state=100, oob_score=True)

In [ ]:
rf.fit(X_train_sm_pca, y_train_sm)

In [ ]:
rf.oob_score_

In [ ]:
from sklearn.metrics import roc_auc_score, RocCurveDisplay
RocCurveDisplay.from_estimator(rf, X_train_sm_pca, y_train_sm, name="ROC Curve")

### Hyper-parameter tuning for the Random Forest

In [ ]:
rf = RandomForestClassifier(random_state=42, n_jobs=-1)

In [ ]:
params = {
    'max_depth': [2,3,5],
    'min_samples_leaf': [50,100],
    'min_samples_split': [ 100, 150, ],
    'n_estimators': [100, 200 ]
}

In [ ]:
grid_search = GridSearchCV(estimator=rf,
                           param_grid=params,
                           cv = 4,
                           n_jobs=-1, verbose=1, scoring="accuracy")

In [ ]:
grid_search.fit(X_train_sm_pca, y_train_sm)

In [ ]:
grid_search.best_score_

In [ ]:
grid_search.best_params_

In [ ]:
rfc_model = RandomForestClassifier(bootstrap=True,
                             max_depth=5,
                             min_samples_leaf=50, 
                             min_samples_split=100,
                             n_estimators=200)

In [ ]:
rfc_model.fit(X_train_sm_pca, y_train_sm)

In [ ]:
evaluate_model(rfc_model)

### Conclusion
- The model accuracy are below, Decision Tree with PCA performs better than other model

### Logistic regression 
- train : ~73%
- test  : ~72%  

### Logistic Regression using PCA 
- train : ~71%
- test  : ~69%

### Decision Tree using PCA 
- train : ~89%
- test  : ~76%

### Random Forest with PCA 
- train : ~75%
- test  : ~71%

### Creating submission file
- For submission, we need to make sure that the format is exactly the same as the sample.csv file. It contains 2 columns, id and churn_probability